### PR TITLE
GE-614: Generate seed data script

### PR DESCRIPTION
**Pull request by ruands**

### CODE REVIEW

File: access_control/db_actions.py [hunk 1]
```diff
@@ -186,3 +186,16 @@ def transform(
...
         data = api_model.from_dict(transformer.apply(data))
     return data
+
+
+def get_or_create(model, **kwargs):
+    """Django-like helper method to get or create objects.
+    """
+    instance = db.session.query(model).filter_by(**kwargs).first()
+    if instance:
+        return instance
```
Comment (AltusBarry) [inline access_control/db_actions.py]: `return instance, False`

File: access_control/db_actions.py [hunk 2]
```diff
@@ -186,3 +186,16 @@ def transform(
...
+    else:
+        instance = model(**kwargs)
+        db.session.add(instance)
+        db.session.commit()
+        return instance
```
Comment (AltusBarry) [inline access_control/db_actions.py]: `return instance, True`

File: access_control/fixtures/load_domains.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
+# Domain hierachy based on:
+# https://praekelt.atlassian.net/wiki/spaces/GEM/pages/215515423/Roles+Permissions+Guidelines?preview=/55574535/118751390/Domain%20Hierarchy%20(5).png
+DOMAIN_HIERARCHY = {
+    "description": "The top level of the organisation",
+    "roles": {},
+    "subdomains": {
+        "Springster": {
```
Comment (AltusBarry) [inline access_control/fixtures/load_domains.py]: should the sub domains be Title case? should they not perhaps all be lower?

File: access_control/fixtures/load_domains.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            "description": "Springster",
+            "roles": {},
+            "subdomains": {
+                "A Markets": {
```
Comment (AltusBarry) [inline access_control/fixtures/load_domains.py]: perhaps lower and underscores instead of spaces?

File: access_control/fixtures/load_roles.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+ROLES = [
+    "ADMIN",
+    "__SUPERUSER__",
+    "SUPER_TECH_ADMIN",
```
Comment (AltusBarry) [inline access_control/fixtures/load_roles.py]: That is quite a role :laughing:

File: seed_data.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# TODO: Rename file to something more appropriate
```
Comment (AltusBarry) [inline seed_data.py]: This still applicable?

Comment (ruands) [reply seed_data.py]: Only if you think it still needs changing.

File: access_control/fixtures/load_roles.py (hunk 5 above)
Comment (ruands) [reply access_control/fixtures/load_roles.py]: I promise I did not make that up. It's actually in our documentation :joy:

File: seed_data.py (hunk 6 above)
Comment (AltusBarry) [reply seed_data.py]: Should this live in root and not `access_control/`?

File: seed_data.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    def load_roles(self):
+        print("Loading roles...")
+        for role in ROLES:
+            instance = get_or_create(Role, label=role)
+            print("Role %s loaded..." % instance.label)
+        print("Done")
+
+    def load_domain(self, name: str, detail: dict, parent: Domain):
+        print("Loading domain %s..." % name)
+        domain = get_or_create(
```
Comment (AltusBarry) [inline seed_data.py]: We will always load on a single domain basis? Not a list of dicts?

Comment (ruands) [reply seed_data.py]: As far as I know, it will be single domain. Starting with Girl Effect Organization as the top level.

Comment (AltusBarry) [reply seed_data.py]: Oh I see, its a possibly recursive method.

File: access_control/fixtures/load_roles.py (hunk 5 above)
Comment (cobusc) [reply access_control/fixtures/load_roles.py]: @ruands Those were the examples in the architecture doc. You have to use the ones defined in Fritz's doc, remember?

File: seed_data.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            parent_id=parent
+        )
+
+        subdomains = detail.get("subdomains", {})
+        for _name, _detail in subdomains.items():
+            self.load_domain(_name, _detail, domain.id)
+
+        roles = detail.get("roles", {})
+        for label, defaults in roles.items():
+            role = Role.query.filter_by(label=label).first_or_404()
```
Comment (AltusBarry) [inline seed_data.py]: Not sure raising a 404 is a good idea here, rather do a try, except with a get and log + raise properly.

File: seed_data.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            Domain, name=name,
+            description=detail["description"],
+            parent_id=parent
+        )
+
+        subdomains = detail.get("subdomains", {})
+        for _name, _detail in subdomains.items():
+            self.load_domain(_name, _detail, domain.id)
+
+        roles = detail.get("roles", {})
```
Comment (AltusBarry) [inline seed_data.py]: Perhaps add the roles before calling the function recursively.

File: seed_data.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        for label, defaults in roles.items():
+            role = Role.query.filter_by(label=label).first_or_404()
+            domain_role = get_or_create(
+                DomainRole, domain=domain, role=role, defaults=defaults)
+
+        print("Done")
+
+
+loader = SeedDataLoader()
+if len(sys.argv) == 2:
```
Comment (AltusBarry) [inline seed_data.py]: When cli args are used, add help text and errors if incorrectly used.

File: seed_data.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    seed = sys.argv[1]
+    func = getattr(loader, "load_%s" % seed)
+    if seed == "domain":
+        func("Girl Effect Organisation", DOMAIN_HIERARCHY, None)
```
Comment (AltusBarry) [inline seed_data.py]: Not sure we should limit in the event its the load domain command. See my list of recursive domains suggestion.

File: access_control/db_actions.py [hunk 12]
```diff
@@ -186,3 +186,16 @@ def transform(
         }
         data = api_model.from_dict(transformer.apply(data))
     return data
+
+
+def get_or_create(model, **kwargs):
```
Comment (AltusBarry) [inline access_control/db_actions.py]: At a test for this one pretty plz. Yes I am sticking with plz! :stuck_out_tongue:

File: seed_data.py (hunk 6 above)
Comment (ruands) [reply seed_data.py]: It's the only place I can get the imports to work properly.

File: seed_data.py (hunk 10 above)
Comment (ruands) [reply seed_data.py]: Removing. Can run all at once.

**Comment by AltusBarry** [comment]
+1, just pick what you wanna do about the TODO and remove it. I have no particular feelings about the naming.
